xosview.1
.TH "XOSVIEW" 1
.UC
.SH NAME
xosview \- X based system monitor
.SH SYNOPSIS
xosview [options]
.\"  ================  Macros  =============================
.\"  First, let's define some handy roff macros.
.\"  A macro begins with .de <xx> where one will invoke this macro with .xx
.\"  The macro definition ends with the .. line.
.\"  I don't know what macro abbreviations are free, so I just chose a few,
.\"  and haven't noticed a problem so far!  bgrayson

.\"  There are several paragraphs that are repeated in the resource section.
.\"  Rather than typing the whole stuff out each time, we define a few macros.

.\"  The .pp macro takes a single argument (net, disk, etc), and
.\" prints a paragraph description of a Priority resource.  Only the
.\" header (xosview*diskPriority: \fIpriority\fP) needs to be specified in
.\" addition to the .pp macro.
.\"  Usage:   .pp load
.de pp
xosview*\\$1Priority: \fIpriority\fP
.RS
This number (which must be an integer >= 1) sets the number of tenths of
a second that the \\$1 meter waits between updates.  A value of 1 has xosview
update the meter 10 times per second (the fastest).  A value of 600 would
cause xosview to update the meter once a minute.
.RE
..
.\"  The .dc macro is similar to the .pp macro, except that it is for
.\" the decay resource paragraphs.
.\"  Usage:  .dc net
.de dc
xosview*\\$1Decay: (True or False)
.RS
If True then the \\$1 meter will be split vertically in two.  The top
half will show the instantaneous state, while the bottom half will
display a decaying average of the state.
.RE
..
.\" The .dg macro is similar to the .dc macro, except that it is for
.\" the scrolling graph resource paragraphs.
.\" Usage:  .dg net
.de dg
xosview*\\$1Graph: (True or False)
.RS
If this is set to True then the \\$1 meter will be drawn as a horizontally
scrolling bargraph showing the state value verses time.
.RE
..
.\"  The .pm macro is for ``plus/minus'' -- for the
.\"  enable/disable command-line options.
.de pm
-\\$1
.RS
This option overrides the xosview*\\$1 resource.  It is
equivalent to setting xosview*\\$1 to "False".
.RE
+\\$1
.RS
This option overrides the xosview*\\$1 resource.  It is
equivalent to setting xosview*\\$1 to "True".
.RE
..
.\"  The .xt macro is for ``Xresource true'' -- for enabling a
.\"  meter, like the .pm macro/-+foo.
.de xt
xosview*\\$1: (True or False)
.RS
If True then xosview will display a \\$1 meter.
.RE
..
.\"  The .uf macro is for the ``used format'' stuff.
.de uf
xosview*\\$1UsedFormat: (percent, scale1000, or scale1024)
.RS
This resource tells xosview how to display "used" labels.  The formats work
as follows:

\fBpercent\fP:
.RS
Display the value as a percentage of the total.
.RE
\fBscale1000\fP:
.RS
Display the absolute value and automatically print the units (K, M, G, etc.) as
appropriate for powers of 1000.
.RE
\fBscale1024\fP:
.RS
Display the absolute value and automatically print the units (K, M, G, etc.) as
appropriate for powers of 1024.
.RE
float:
.RS
Deprecated: Use 'scale1000' instead.
.RE
autoscale:
.RS
Deprecated: Use 'scale1024' instead.
.RE
.RE
..
.\"  Define a color macro for the various xosview*fooXXXColor: resources.
.\"  Usage:  .cc swap Used used "swap space"  <--  keep 'swap
\"                                                       space' as one arg.
.de cc
xosview*\\$1\\$2Color: \fIcolor\fP
.RS
The \\$1 meter will use this color to display the \\$3 field.
.RE
..
.\"  ================  End of Macros  =============================
.SH DESCRIPTION
\fIxosview\fP is a monitor which displays the status of several system
parameters.  These include CPU usage, load average, memory, swap
space, network usage and more.  Each resource is displayed as a
horizontal bar which is separated into color coded regions showing how
much of the resource is being put to a particular use.

\fIxosview\fP runs on several platforms (Linux, NetBSD, FreeBSD,
OpenBSD, DragonflyBSD, some Solaris systems, IRIX 6.5, HPUX and GNU).
Not all of the meters described below are supported on all platforms.
Some of the meters may appear different depending upon the platform
\fIxosview\fP is running on.  Note that *BSD is used as an
abbreviation for all the supported BSD operating systems (NetBSD,
FreeBSD, OpenBSD, DragonflyBSD).

\fBLoad\fP: Kernel reported load average
   all platforms  :  load
   Linux          :  load | CPU frequency
   *BSD           :  load | CPU frequency

\fBCPU Usage\fP
   Linux  : user | nice | system | soft-interrupt | interrupt | io-wait |
            guest | niced guest | stolen | idle
   *BSD   : user | nice | system | interrupt | idle
   IRIX   : user | system | interrupt | wait | idle
   HPUX   : user | nice | system | interrupt | idle

\fBMemory Usage\fP
   Linux  : used | buffers | slab | map | cache | free
   FreeBSD: active | inactive | wired | cache | free
   DFBSD  : active | inactive | wired | cache | free
   OpenBSD: active | inactive | wired | free
   NetBSD : active | inactive | wired | free
   IRIX   : kernel | fs | user | free
   HPUX   : text | used | other | free
   GNU    : active | inactive | wired | free

\fBSwap Usage\fP
   Linux  : used | free
   *BSD   : used | free
   HPUX   : used | free
   GNU    : used | free

\fBDisk Usage\fP: Throughput to/from local disk, per second
   Linux  : in | out | idle
   *BSD   : in | out | idle

\fBPage Swapping\fP: Pages to/from swap, per second
   Linux  : in | out | idle
   *BSD   : in | out | idle
   HPUX   : in | out | idle
   GNU    : in | out | idle

\fBNetwork Usage\fP
   Linux  : in | out | idle
   *BSD   : in | out | idle

\fBGFX Usage\fP: Framebuffer exchanges, per second
   IRIX   : swapbuffers

\fBInterrupts\fP: "leds" which blink when an interrupt occurs
   Linux  : IRQs
   *BSD   : IRQs

\fBInterrupt Rate\fP: Per second
   *BSD   : interrupts | free

\fBSerial Port Status\fP: "leds" which show the serial port parameters
   Linux  : LSR and MSR

\fBBattery Level\fP: charge and status of batteries
   Linux  : available | used

\fBRAID\fP: State of disks in a software RAID array
   Linux  : disk0 disk1 disk2 ... diskN | rebuild

\fBWireless Link\fP
   Linux  : quality

\fBTemperature sensors\fP
   Shows readings from temperature sensors.
   Linux  : ACPI or i2c/hwmon or Intel Core/AMD K8+ sensors
   *BSD   : i2c or Intel Core/AMD K8+ sensors

Typing a 'q' in the window will terminate xosview.

.SH OPTIONS

Most of these command line options are just a convenient way to set one or
more of \fIxosview\fP's resources.  Please see the \fBRESOURCES\fP section for
more details on what the resource is for.

\-v
.RS
Displays the version number.
.RE

\-name \fIname\fP
.RS
Sets the Resource name xosview will use (same as the \-name option
for typical X applications like xterm).  When performing resource
lookups, xosview will look under \fIname\fP, and then under
``xosview''.  For an example, perform the following (as
documented in README.netbsd):
.RS
  xrdb \-merge Xdefaults.stipple
  xosview \-name xosvstipple &
  xosview \-name xosvstipplebw &
.RE
.RE

\-display \fIdisplay\fP
.RS
Sets the X display to \fIdisplay\fP.  This option overrides the
xosview*display resource.
.RE

\-font \fIfont\fP
.RS
Specifies the font xosview will use for labels.  This option overrides the
xosview*font resource.
.RE

\-title \fItitle\fP
.RS
This option sets the name xosview will tell the window manager to use for
the X window and icon.  This option overrides the xosview*title resource.
.RE

\-geometry \fIgeometry_string\fP
.RS
Sets the X geometry to \fIgeometry_string\fP.  This option overrides the
xosview*geometry resource.
.RE

.\"  Handle -/+captions
.pm captions

.\"  Handle -/+labels
.pm labels

\-usedlabels
.RS
This option overrides the xosview*usedlabels resource.  It is equivalent to
setting xosview*usedlabels to "False".
.RE
+usedlabels
.RS
This option overrides the xosview*usedlabels resource.  It is equivalent to
setting xosview*usedlabels to "True".
.RE

.\"  Handle -/+cpu
.pm cpu

.\"  Handle -/+load
.pm load

.\"  Handle -/+mem
.pm mem

.\"  Handle -/+swap
.pm swap

.\"  Handle -/+battery
.pm battery

.\"  Handle -/+gfx
.pm gfx

.\"  Handle -/+wireless
.pm wireless

.\"  Handle -/+net
.pm net

\-network \fImaxbandwidth\fP
.RE
\-networkBW \fImaxbandwidth\fP
.RE
\-networkBandwidth \fImaxbandwidth\fP
.RS
These options override the xosview*netBandwidth resource.  They cause
xosview to display a meter that will shows network usage, with a maximum
bandwidth of \fBmaxbandwidth\fP.  Notice that setting the bandwidth to
0 no longer disables the meter \-\- use the ``\-net'' option instead.
.RE

.\"  Handle \-/+page
.pm page

\-pagespeed \fIval\fP
.RS
This option overrides the xosview*pageBandWidth resource.  The resource
xosview*pageBandWidth will be set to \fIval\fP.
.RE

.\"  Handle -/+disk
.pm disk

.\"  Handle -/+ int
.pm int

\-ints +ints
.RE
\-interrupts +interrupts
.RS
Equivalent to \-int and +int.
.RE

.pm irqrate

\-intrate +intrate
.RS
Equivalent to \-irqrate and +irqrate.
.RE

.\"  Handle -/+lmstemp
.pm lmstemp

.\"  Handle -/+coretemp
.pm coretemp

.\"  Handle -/+acpitemp
.pm acpitemp

.\"  Handle -/+bsdsensor
.pm bsdsensor
.RE

\-xrm \fIresource_string\fP
.RS
This switch allows any of xosview's resources to be set on the command line.
An example of how the xosview*memFreeColor could be set using this option is
shown below (Note the use of " to prevent the shell from expanding
\'*\' or from creating two separate arguments, \'xosview*memfreeColor:\'
and \'purple\'):
.RS
\-xrm "xosview*memFreeColor: purple"
.RE
.RE

.SH X RESOURCES

The following is a list of X resources supported by \fIxosview\fP.  Each has
a default value assigned to it.  These values can be found in the file
Xdefaults which can be obtained in the source distribution of xosview.
They can be overridden in the usual places (/usr/lib/X11/app-defaults/XOsview,
$HOME/.Xdefaults, etc.).

It should be noted that it is OK to have a resource defined for a port of
xosview that does not support the feature the resource configures.  Xosview
will simply ignore the resources that are set for it but not supported on
a given platform.


\fBGeneral Resources\fP

xosview*title: \fIname\fP
.RS
The string that xosview will use for the X window title.  Normally xosview
will use 'xosview@machine_name' for a title.  This resource overrides the
default behavior.
.RE

xosview*geometry: \fIgeometry_string\fP
.RS
This is a standard X geometry string that defines the size and location of
the X window used by xosview.
.RE

xosview*display: \fIname\fP
.RS
The name of the display where xosview will contact the X server for drawing
its window.
.RE

xosview*pixmapName: \fIname\fP
.RS
The filename of an X pixmap (xpm) file for use as a background
image.
.RE

xosview*captions: (True or False)
.RS
If True then xosview will display meter captions.
.RE

xosview*labels: (True or False)
.RS
If True then xosview will display meter labels.
.RE

xosview*meterLabelColor: \fIcolor\fP
.RS
The color to use for the meter labels.
.RE

xosview*usedlabels: (True or False)
.RS
If True then xosview will display labels that show the percentage of the
resource (or absolute amount, depending on the meter) being used.  This
option requires that the labels option also be set to True.
.RE

xosview*usedLabelColor: \fIcolor\fP
.RS
The color to use for "used" labels.
.RE

xosview*borderwidth: \fIwidth\fP
.RS
The width of the border for the xosview window.
.RE

xosview*font: \fIfont\fP
.RS
This is the font that xosview will use.
.RE

xosview*background: \fIcolor\fP
.RS
This is the color that will be used for the background.
.RE

xosview*foreground: \fIcolor\fP
.RS
This is the color that will be used for the foreground.
.RE

xosview*BadDataColor: \fIcolor\fP
.RS
This is the color that will be used indicate that there is no data to
graph.
.RE

xosview*enableStipple: (True or False)
.RS
Change to true to try stipple support.  This is primarily for users
stuck with 1-bit monitors/display cards.  Try setting enableStipple
true.  Please give us feedback on this, if you use it.  It needs
some more work, but no one has given us any feedback so far.
.RE


\fBO/S Type Specific General Resources\fP
.RS

\fBLinux\fP

xosview*linux.metrics: \fImetric list\fP
.RS
Ordered list of metrics to display.
.RE
.RE

\fBLoad Meter Resources\fP

.xt load

.pp load

.dc load

.dg load

.uf load

xosview*loadCpuSpeed:   (True or False)
.RS
Display the current CPU speed in the load meter.
.RE

.\"  loadIdleColor
.cc load Idle idle

xosview*loadColorList: \fIcolor list\fP
.RS
List of colors to be used for the load value color. Colors earlier in the list
are used to display lower load values than colors later in the list.
.RE

xosview*loadScalingAdjust: Auto or \fIfloat\fP
.RS
The system suppiled load average is adjusted by dividing by this value before
the graph is drawn. If it is set to "auto", the number of CPUs is used. If the
value is less than 0.0, the system supplied load average is adjusted by dividing
by the negative of this value time the CPU count.
Default is "Auto".
.RE

xosview*loadScalingBase: \fIfloat\fP
.RS
The "base" for the exponential multiplier. It must be greater than 1.0; 2.0
is good choice.
Default is 4 to match the color behavior of older versions.
.RE

xosview*loadScalingBehavior: \fIbehavior\fP
.RS
This determines how the Load Meter displays the load value.
Default is "traditional".

\fBTraditional\fP
.RS
This behavior is similar to that of older versions of \fIxosview\fP. The color
of the load value color indicates the adjusted scale range of the graph. The
first color has an adjusted range of [0.0, 1.0]; the second color is
[0, Base]; the third color is [0, Base^2]; etc. If the adjusted load value
exceeds the scale range of the last color, the meter will "peg".
.RE

\fBOverlay\fP
.RS
With this behavior, the graphed load value appears to grow from the low end
of the scale as the displayed load increases; overlaying the previous display.
The color of the load value color indicates the adjusted scale range of the
graph. The first color has an adjusted range of [0.0, 1.0]; the second color
is (1.0, Base]; the third color is (Base, Base^2]; etc. If the adjusted load
value exceeds the scale range of the last color, the meter will "peg".
.RE
.RE


\fBCPU Meter Resources\fP

xosview*cpu: (True or False)
.RS
If True then xosview will display a cpu meter.  On Linux and IRIX SMP
machines, the resource cpuFormat defines how meters are created for
multiple CPUs.
.RE

.\"  cpuUserColor, cpuNiceColor, cpuSystemColor, cpuInterruptColor, cpuFreeColor
.cc cpu User "cpu user time"

.cc cpu Nice "cpu nice time"

.cc cpu System "cpu system time"

.cc cpu Interrupt "cpu hard interrupt time"

.cc cpu SInterrupt "cpu soft interrupt time"

.cc cpu Wait "cpu IO waiting time"

.cc cpu Guest "cpu virtualization guest time"

.cc cpu NiceGuest "cpu niced virtualization guest time"

.cc cpu Stolen "cpu involuntary wait time"

.cc cpu Free "cpu idle time"

.\"  Priority, decay, usedFormat resources:
.pp cpu

.dc cpu

.dg cpu

.uf cpu

xosview*cpuFormat: (single, all, both or auto)
.RS
If `single', only a cumulative meter for all CPU usage is created.
`all' creates a meter for each CPU, but no cumulative meter. `both'
creates one cumulative meter and one for each CPU. `auto' makes a
choice based on the number of CPUs found.
.RE

xosview*cpuFields: USED/USR/NIC/SYS/INT/SI/HI/WIO/GST/NGS/STL/IDLE
.RS
The set of fields to show in Linux CPU meter instead of the
default. Possible fields are:

\fBUSED\fP:
.RS
Combine all used CPU time into one field. This is the sum of user, nice,
system, soft and hard interrupts, guest, niced guest and stolen times.
None of these, except stolen, may be defined together with `USED'.
.RE
\fBIDLE\fP:
.RS
Time spent doing nothing. Includes I/O wait if it is not defined separately.
.RE
\fBUSR\fP:
.RS
Time spent in user mode processes. Includes nice, guest and niced guest if
those are not defined separately.
.RE
\fBNIC\fP:
.RS
Time spent in niced user mode processes. Includes niced guest if neither it
nor guest is not defined separately.
.RE
\fBSYS\fP:
.RS
Time spent in kernel code. Includes soft and hard interrupt as well as stolen
time if those are not defined separately.
.RE
\fBINT\fP:
.RS
Combines soft and hard interrupt handling times into one field.
.RE
\fBSI\fP:
.RS
Time the kernel used to handle soft interrupts. Available on Linux kernel
2.6.0 and higher.
.RE
\fBHI\fP:
.RS
Time the kernel used to handle hard interrupts. Available on Linux kernel
2.6.0 and higher.
.RE
\fBWIO\fP:
.RS
Time spent waiting for I/O to complete. Available on Linux kernel
2.6.0 and higher.
.RE
\fBGST\fP:
.RS
Time spent running guest OS in virtual machine. Includes niced guest if
it is not defined separately. Available on Linux kernel 2.6.24 and higher.
.RE
\fBNGS\fP:
.RS
Time spent running niced guest OS in virtual machine. Available on Linux kernel
2.6.32 and higher.
.RE
\fBSTL\fP:
.RS
Involuntary wait time when running as guest in virtual machine. Available
on Linux kernel 2.6.11 and higher.
.RE

Most combinations are possible (see above for restrictions), but at least
`USED' or `USR' and `SYS' need to be defined. `IDLE' field is added
automatically.
.RE


\fBMemory Meter Resources\fP

.\"  Do the mem: True resource.
.xt mem

.\"  mem{Used,Share,Buffer,...}Color resources
.cc mem Used "used memory"

.cc mem Shared "shared memory"

.cc mem Buffer "buffer memory"

.cc mem Cache "cache memory"

.cc mem Free "free memory"

.cc mem Shared "shared memory"

.cc mem Text "HP text memory"

.cc mem Other "HP ``other'' memory"

.cc mem Active "*BSD active memory"

.cc mem Inactive "*BSD inactive memory"

.cc mem Wired "*BSD wired memory"

.cc mem Slab "Linux in-kernel data structures"

.cc mem Map "Linux memory mapped files"

.\"  Priority, decay, usedFormat resources:
.pp mem

.dc mem

.dg mem

.uf mem


\fBSwap Meter Resources\fP

.\"  Do the swap: True resource.
.xt swap

.\"  swap{Used,Free}Color resources.
.cc swap Used "used swap"

.cc swap Free "free swap"

.\"  Priority, decay, usedFormat resources:
.pp swap

.dc swap

.dg swap

.uf swap


\fBPage Swapping Meter Resources\fP

.\"  Do the page: True resource.
.xt page

xosview*pageBandWidth: \fImaxEvents\fP
.RS
This number is used to specify the expected maximum bandwidth (in events /
sec) for the page meter.  When the expected maximum bandwidth
(\fImaxEvents\fP) is exceeded then the page meter will display the relative
percentage of page swapping (25% in, 75% out).
.RE

.\"  page{In,Out,Idle}Color:
.cc page In page-in

.cc page Out page-out

.cc page Idle idle

.\"  Priority, decay, usedFormat resources:
.pp page

.dc page

.dg page

.uf page


\fBGfx Meter Resources\fP

xosview*gfx:	(True or False)
.RS
If True xosview will display the GfxMeter. The value is sampled once per
second, due to the usage of sadc to sample data.
.RE

xosview*gfxWarnColor: \fIcolor\fP
.RS
This is the color that the gfx meter will use once the warn state is
reached.
.RE

xosview*gfxAlarmColor: \fIcolor\fP
.RS
This is the color that the gfx meter will use once the alarm state is
reached.
.RE

xosview*gfxSwapColor: \fIcolor\fP
.RS
This is the color that the gfx meter will use in normal state
.RE

.\"  gfxIdleColor
.cc gfx Idle idle

.\"  Do the priority resource
.pp gfx

xosview*gfxWarnThreshold: \fIint\fP
.RS
This number (which must be an integer >= 1) of swapbuffers per second
and pipe at which the gfxmeter changes its status and color from
"normal" to "warn". The default value is 60.
.RE

xosview*gfxAlarmThreshold: \fIint\fP
.RS
This number (which must be an integer >= gfxWarnThreshold) of
swapbuffers per second and pipe at which the gfxmeter changes its status
and color from "warn" to "alarm". The default value is 120.
.RE


xosview.cc
//
//  Copyright (c) 1994, 1995, 2002, 2006 by Mike Romberg ( mike.romberg@noaa.gov )
//
//  This file may be distributed under terms of the GPL
//

#include <unistd.h>
#include <stdlib.h>
#include <sys/time.h>
#include "xosview.h"
#include "meter.h"
#include "MeterMaker.h"
#if ( defined(XOSVIEW_NETBSD) || defined(XOSVIEW_FREEBSD) || \
      defined(XOSVIEW_OPENBSD) || defined(XOSVIEW_DFBSD) )
# include "kernel.h"
#endif

static const char * const versionString = "xosview version: Git";

static const char NAME[] = "xosview@";

#if !defined(__GNUC__)

#define MIN(x,y)		\
(				\
    x < y ? x : y		\
)

#define MAX(x,y)		\
(				\
    x > y ? x : y		\
)

#else

#define MIN(x,y)		\
({				\
    const typeof(x) _x = x;	\
    const typeof(y) _y = y;	\
				\
    (void) (&_x == &_y);	\
				\
    _x < _y ? _x : _y;		\
})

#define MAX(x,y)		\
({				\
    const typeof(x) _x = x;	\
    const typeof(y) _y = y;	\
				\
    (void) (&_x == &_y);	\
				\
    _x > _y ? _x : _y;		\
})

#endif // sgi

double MAX_SAMPLES_PER_SECOND = 10;

XOSView::XOSView( const char * instName, int argc, char *argv[] ) : XWin(),
						xrm(Xrm("XOSView", instName)){
  // Check for version arguments first.  This allows
  // them to work without the need for a connection
  // to the X server
  checkVersion(argc, argv);

  setDisplayName (xrm.getDisplayName( argc, argv));
  openDisplay();  //  So that the Xrm class can contact the display for its
		  //  default values.
  //  The resources need to be initialized before calling XWinInit, because
  //  XWinInit looks at the geometry resource for its geometry.  BCG
  xrm.loadAndMergeResources (argc, argv, display_);
  XWinInit (argc, argv, NULL, &xrm);
#if 1	//  Don't enable this yet.
  MAX_SAMPLES_PER_SECOND = atof(getResource("samplesPerSec"));
  if (!MAX_SAMPLES_PER_SECOND)
    MAX_SAMPLES_PER_SECOND = 10;
#endif
  usleeptime_ = (unsigned long) (1000000/MAX_SAMPLES_PER_SECOND);
  if (usleeptime_ >= 1000000) {
    /*  The syscall usleep() only takes times less than 1 sec, so
     *  split into a sleep time and a usleep time if needed.  */
    sleeptime_ = usleeptime_ / 1000000;
    usleeptime_ = usleeptime_ % 1000000;
  } else { sleeptime_ = 0; }
#if ( defined(XOSVIEW_NETBSD) || defined(XOSVIEW_FREEBSD) || \
      defined(XOSVIEW_OPENBSD) || defined(XOSVIEW_DFBSD) )
  BSDInit();	/*  Needs to be done before processing of -N option.  */
#endif

  hmargin_  = atoi(getResource("horizontalMargin"));
  vmargin_  = atoi(getResource("verticalMargin"));
  vspacing_ = atoi(getResource("verticalSpacing"));
  hmargin_  = MAX(0, hmargin_);
  vmargin_  = MAX(0, vmargin_);
  vspacing_ = MAX(-BORDER_WIDTH, vspacing_);

  checkArgs (argc, argv);  //  Check for any other unhandled args.
  xoff_ = hmargin_;
  yoff_ = 0;
  nummeters_ = 0;
  meters_ = NULL;
  name_ = const_cast<char *>("xosview");
  _deferred_resize = true;
  _deferred_redraw = true;
  windowVisibility = OBSCURED;

  //  set up the X events
  addEvent( new Event( this, ConfigureNotify,
		      (EventCallBack)&XOSView::resizeEvent ) );
  addEvent( new Event( this, Expose,
		      (EventCallBack)&XOSView::exposeEvent ) );
  addEvent( new Event( this, KeyPress,
		      (EventCallBack)&XOSView::keyPressEvent ) );
  addEvent( new Event( this, VisibilityNotify,
                      (EventCallBack)&XOSView::visibilityEvent ) );
  addEvent( new Event( this, UnmapNotify,
                      (EventCallBack)&XOSView::unmapEvent ) );

  // add or change the Resources
  MeterMaker mm(this);

  // see if legends are to be used
  checkOverallResources ();

  // add in the meters
  mm.makeMeters();
  for (int i = 1 ; i <= mm.n() ; i++)
    addmeter(mm[i]);

  if (nummeters_ == 0)
  {
    fprintf (stderr, "No meters were enabled!  Exiting...\n");
    exit (0);
  }

  //  Have the meters re-check the resources.
  checkMeterResources();

  // determine the width and height of the window then create it
  figureSize();
  init( argc, argv );
  title( winname() );
  iconname( winname() );
  dolegends();
  sampleClock = 0;
}


XOSView::~XOSView( void ){
  MeterNode *tmp = meters_;
  while ( tmp != NULL ){
    MeterNode *save = tmp->next_;
    delete tmp->meter_;
    delete tmp;
    tmp = save;
  }
}


void XOSView::checkVersion(int argc, char *argv[]) const
    {
    for (int i = 0 ; i < argc ; i++)
        if (!strncasecmp(argv[i], "-v", 2)
          || !strncasecmp(argv[i], "--version", 10))
            {
            std::cerr << versionString << std::endl;
            exit(0);
            }
    }

void XOSView::figureSize(void) {
  if ( legend_ ){
    if ( !usedlabels_ )
      xoff_ += textWidth( "XXXXX" );
    else
      xoff_ += textWidth( "XXXXXXXXX" );

    yoff_ = caption_ ? textHeight() : 0;
  }
  static int firsttime = 1;
  if (firsttime) {
    firsttime = 0;
    width_ = 2*hmargin_ + textWidth( "XXXXXXXXXXXXXXXXXXXXXXXXXXXXX" );
    height_ = 2*vmargin_ + nummeters_*(vspacing_+yoff_+textHeight()) - vspacing_;
  }
}

void XOSView::checkMeterResources( void ){
  MeterNode *tmp = meters_;

  while ( tmp != NULL ){
    tmp->meter_->checkResources();
    tmp = tmp->next_;
  }
}

int XOSView::newypos( void ){
  return 15 + 25 * nummeters_;
}

void XOSView::dolegends( void ){
  MeterNode *tmp = meters_;
  while ( tmp != NULL ){
    tmp->meter_->docaptions( caption_ );
    tmp->meter_->dolegends( legend_ );
    tmp->meter_->dousedlegends( usedlabels_ );
    tmp = tmp->next_;
  }
}

void XOSView::addmeter( Meter *fm ){
  MeterNode *tmp = meters_;

  if ( meters_ == NULL )
    meters_ = new MeterNode( fm );
  else {
    while ( tmp->next_ != NULL )
      tmp = tmp->next_;
    tmp->next_ = new MeterNode( fm );
  }
  nummeters_++;
}

void XOSView::checkOverallResources() {
  //  Check various resource values.

  //  Set 'off' value.  This is not necessarily a default value --
  //    the value in the defaultXResourceString is the default value.
  usedlabels_ = legend_ = caption_ = 0;

  setFont();

   // use captions
  if ( isResourceTrue("captions") )
      caption_ = 1;

  // use labels
  if ( isResourceTrue("labels") )
      legend_ = 1;

  // use "free" labels
  if ( isResourceTrue("usedlabels") )
      usedlabels_ = 1;
}

const char *XOSView::winname( void ){
  char host[100];
  gethostname( host, 99 );
  static char name[101];	/*  We return a pointer to this,
				    so it can't be local.  */
  snprintf( name, 100, "%s%s", NAME, host);
  //  Allow overriding of this name through the -title option.
  return getResourceOrUseDefault ("title", name);
}


void XOSView::collect(void) {
  MeterNode *tmp = meters_;

  XOSDEBUG("Doing collect.\n");

  while (tmp != NULL) {
    if (tmp->meter_->requestevent(sampleClock))
      tmp->meter_->checkevent();
    tmp = tmp->next_;
  }
}


void XOSView::history(void) {
  MeterNode *tmp = meters_;

  XOSDEBUG("Doing history.\n");

  while (tmp != NULL) {
    if (tmp->meter_->requestevent(sampleClock))
      tmp->meter_->updateMeterHistory();
    tmp = tmp->next_;
  }
}


void  XOSView::resize(void) {
  int newwidth;
  int newheight;
  int vremain;

  XOSDEBUG("Doing resize.\n");

  newwidth = width_ - xoff_ - hmargin_;
  newwidth = (newwidth > 2*BORDER_WIDTH+1) ? newwidth : 2*BORDER_WIDTH+1;

  newheight = (height_ - (2*vmargin_ + nummeters_*(vspacing_+yoff_) - vspacing_))
      / nummeters_;
  newheight = (newheight > 2*BORDER_WIDTH+1) ? newheight : 2*BORDER_WIDTH+1;

  vremain = (height_ - 2*vmargin_ - nummeters_*(vspacing_+yoff_+newheight) + vspacing_)/2;
  vremain = (vremain > 0) ? vremain : 0;

  int counter = 0;
  MeterNode *tmp = meters_;
  while ( tmp != NULL ) {
    tmp->meter_->resize( xoff_, vmargin_+yoff_+vremain + counter*(newheight+vspacing_+yoff_),
                        newwidth, newheight );
    tmp = tmp->next_;

    counter++;
  }
}


void XOSView::draw(void) {
  MeterNode *tmp = meters_;

  XOSDEBUG("Doing draw.\n");

  clear();

  while (tmp != NULL) {
    tmp->meter_->drawMeterDisplay();
    tmp = tmp->next_;
  }
}


void XOSView::update(void) {
  MeterNode *tmp = meters_;

  XOSDEBUG("Doing update.\n");

  while (tmp != NULL) {
    if (tmp->meter_->requestevent(sampleClock))
      tmp->meter_->updateMeterDisplay();
    tmp = tmp->next_;
  }
}


void XOSView::flushX(bool force) {
  if (!force) {
    MeterNode *tmp = meters_;

    while ( tmp != NULL ){
      if (tmp->meter_->requestevent(sampleClock)) {
        force = true;
        break;
      }
      tmp = tmp->next_;
    }
  }

  if (force) {
    XOSDEBUG("Doing flush.\n");

    flush();
  }
}


void XOSView::run(void) {
  while(!done_) {
    // Update the metrics
    collect();

    // Update meter histories
    // FIXME: This is separate from collect() to reduce sample jitter.
    history();

    // Check for X11 events
    checkevent();

    // Check if the window has been resized (at least once)
    if (_deferred_resize) {
      resize();
      _deferred_resize = false;
      _deferred_redraw = true;
    }

    if (windowVisibility == OBSCURED) {
      _deferred_redraw = false;
    }

    // redraw everything if needed
    if (_deferred_redraw) {
      draw();
    } else {
      // just the updated meters
      update();
    }

    flushX(_deferred_redraw);

    _deferred_redraw = false;

    /*  First, sleep for the proper integral number of seconds --
     *  usleep only deals with times less than 1 sec.  */
    if (sleeptime_) sleep((unsigned int)sleeptime_);
    if (usleeptime_) usleep( (unsigned int)usleeptime_);

    sampleClock++;
  }
}


void XOSView::keyPressEvent( XKeyEvent &event ){
  char c = 0;
  KeySym key;

  XLookupString( &event, &c, 1, &key, NULL );

  if ( (c == 'q') || (c == 'Q') )
    done_ = 1;
}

void XOSView::checkArgs (int argc, char** argv) const
{
  //  The XWin constructor call in the XOSView constructor above
  //  modifies argc and argv, so by this
  //  point, all XResource arguments should be removed.  Since we currently
  //  don't have any other command-line arguments, perform a check here
  //  to make sure we don't get any more.
  if (argc == 1) return;  //  No arguments besides X resources.

  //  Skip to the first real argument.
  argc--;
  argv++;
  while (argc > 0 && argv && *argv)
  {
    switch (argv[0][1]) {
      case 'n': //  Check for -name option that was already parsed
		//  and acted upon by main().
		if (!strncasecmp(*argv, "-name", 6))
		{
		  argv++;	//  Skip arg to -name.
		  argc--;
		}
		break;
#if ( defined(XOSVIEW_NETBSD) || defined(XOSVIEW_FREEBSD) || \
      defined(XOSVIEW_OPENBSD) || defined(XOSVIEW_DFBSD) )
      case 'N': if (strlen(argv[0]) > 2)
		  SetKernelName(argv[0]+2);
		else
		{
		  SetKernelName(argv[1]);
		  argc--;
		  argv++;
		}
		break;
#endif
	      /*  Fall through to default/error case.  */
      default:
		std::cerr << "Ignoring unknown option '" << argv[0] << "'.\n";
		break;
    }
    argc--;
    argv++;
  }
}


void XOSView::exposeEvent(XExposeEvent &event) {
  int i = event.type;	// to quiet the compiler
  i = i;

  _deferred_redraw = true;
  XOSDEBUG("Got expose event.\n");
}


/*
 * All window changes come in via XConfigureEvent (not
 * XResizeRequestEvent)
 */
void XOSView::resizeEvent(XConfigureEvent &event) {
  XOSDEBUG("Got configure event.\n");

  if (event.width == width_ && event.height == height_)
    return;

  XOSDEBUG("Window has resized\n");

  width(event.width);
  height(event.height);
  _deferred_resize = true;
}


void XOSView::visibilityEvent(XVisibilityEvent &event) {
  if (event.state == VisibilityPartiallyObscured) {
    if (windowVisibility != FULLY_VISIBLE)
      _deferred_redraw = true;
    windowVisibility = PARTIALLY_VISIBILE;
  }
  else if (event.state == VisibilityFullyObscured) {
    windowVisibility = OBSCURED;
    _deferred_redraw = false;
  }
  else {
    if (windowVisibility != FULLY_VISIBLE)
      _deferred_redraw = true;
    windowVisibility = FULLY_VISIBLE;
  }

  XOSDEBUG("Got visibility event: %s\n",
    (windowVisibility == FULLY_VISIBLE)
        ? "Full"
        : (windowVisibility == PARTIALLY_VISIBILE)
            ? "Partial"
            : "Obscured"
    );
}


void XOSView::unmapEvent(XUnmapEvent & ev) {
  /* unclutter creates a subwindow of our window if it hides the cursor,
     we get the unmap event if the cursor is moved again. Don't treat it
     as main window unmap */
  if(ev.window == window_)
    windowVisibility = OBSCURED;
}
